### ISSUE TITLE
Error fcopy: invalid ownership for file /etc/resolv.conf (systemd-resolved)

### ISSUE THREAD

**Issue by crocket** [netblue30/firejail#4545]
I want to run a sandboxed program with

```
sudo -u xxx firejail --private-etc=file prog
```

`--private-etc=file` results in

```
Error fcopy: invalid ownership for file /etc/file
```

because

```
$ ls -lh /etc/file
-rw-r----- 1 xxx xxx ... /etc/file
```

Let's say owner and group are `xxx` which is neither root nor a regular user that runs my desktop environment.

A workaround is

```
--whitelist=/etc/file
```
Version
```
$ firejail --version
firejail version 0.9.66

Compile time support:
        - always force nonewprivs support is disabled
        - AppArmor support is enabled
        - AppImage support is enabled
        - chroot support is enabled
        - D-BUS proxy support is enabled
        - file and directory whitelisting support is enabled
        - file transfer support is enabled
        - firetunnel support is disabled
        - networking support is enabled
        - output logging is enabled
        - overlayfs support is disabled
        - private-home support is enabled
        - private-cache and tmpfs as user enabled
        - SELinux support is disabled
        - user namespace support is enabled
        - X11 sandboxing support is enabled
```

OS: Gentoo Linux

Relates to:

* #1531

**Comment by kmk3** [comment]
> I don't know what changed, but my `signal-desktop` suddenly is also broken:

> ```
> Error fcopy: invalid ownership for file /etc/resolv.conf
> Error: failed to run /run/firejail/lib/fcopy, exiting...
> Error: proc 115237 cannot sync with peer: unexpected EOF
> Peer 115241 unexpectedly exited with status 1
> ```
> 
> In `signal-desktop.profile`, there is
> 
> > private-etc alternatives,ca-certificates,crypto-policies,fonts,ld.so.cache,ld.so.conf,ld.so.conf.d,ld.so.preload,localtime,machine-id,nsswitch.conf,pki,resolv.conf,ssl
> 
> Temporary fix is to add `ignore private-etc` in
> `~/.config/firejail/signal-desktop.local`.

What is the firejail version and distribution name/version?

What is the output of `ls -l /etc/resolv.conf`?

Does it happen with firejail-git?

**Comment by kmille** [comment]
I'm running `firejail version 0.9.72` on Arch-Linux. By the way the same happened for Nextcloud client.
```
kmille@linbox:~ ls -l /etc/resolv.conf 
-rw-r--r-- 1 systemd-resolve systemd-resolve 56 Nov 26 11:53 /etc/resolv.conf
```

**Comment by kmille** [comment]
Hmm. The problem came up after the latest update:

```
kmille@linbox:~ firejail --version
firejail version 0.9.73

kmille@linbox: ping -c 1 1.1.1.1
Error fcopy: invalid ownership for file /etc/resolv.conf
Error: failed to run /run/firejail/lib/fcopy, exiting...
Error: proc 62538 cannot sync with peer: unexpected EOF
Peer 62540 unexpectedly exited with status 1
```
Thunderbird and signal also broke. Adding `ignore private-etc` to `globals.local` fixes the problem.

**Comment by kmk3** [comment]
> Hmm. The problem came up after the latest update:

Update of what?

What is the version before and after?

> ```
> kmille@linbox:~ firejail --version
> firejail version 0.9.73
> 
> kmille@linbox: ping -c 1 1.1.1.1
> Error fcopy: invalid ownership for file /etc/resolv.conf
> Error: failed to run /run/firejail/lib/fcopy, exiting...
> Error: proc 62538 cannot sync with peer: unexpected EOF
> Peer 62540 unexpectedly exited with status 1
> 
> kmille@linbox: ls -l /etc/resolv.conf
> -rw-r--r-- 1 systemd-resolve systemd-resolve 63 Nov 12 19:52 /etc/resolv.conf
> ```
> 
> Thunderbird and signal also broke. Adding `ignore private-etc` to
> `globals.local` fixes the problem.

It appears that the code expects /etc/resolv.conf to be a symlink to
/run/systemd/resolve/resolv.conf.

Does it work with the following patch?

```diff
diff --git a/src/fcopy/main.c b/src/fcopy/main.c
index 24f5fd629..61cac877d 100644
--- a/src/fcopy/main.c
+++ b/src/fcopy/main.c
@@ -328,7 +328,9 @@ static char *check(const char *src) {
        //    /run/systemd/resolve/resolv.conf; this file is owned by systemd-resolve user
        // checking gid will fail for files with a larger group such as /usr/bin/mutt_dotlock
        uid_t user = getuid();
-       if (user == 0 && strncmp(rsrc, "/run/systemd/resolve/", 21) == 0) {
+       if (user == 0 && (
+           strncmp(rsrc, "/run/systemd/resolve/", 21) == 0) ||
+           strncmp(rsrc, "/etc/resolv.conf", 16) == 0) {
                // check user systemd-resolve
                struct passwd *p = getpwnam("systemd-resolve");
                if (!p)
```

**Comment by kmille** [comment]
```
[2024-11-05T09:31:01+0100] [ALPM] upgraded firejail-git (0.9.72.r1140.gb2be4870d-1 -> 0.9.72.r1144.gd763fb73c-1)
```
I will check the patch later, thanks!

**Comment by kmille** [comment]
The patch works, thanks! Should we close this?

**Comment by kmk3** [comment]
> The patch works, thanks!

Thanks for testing.

> Should we close this?

I think I'll close it with the patch.

> UPDATE: Btw: this was actually a misconfiguration... Fixed it (but unrelated)

Could you clarify?

Having it as a normal file seems like a valid configuration.

From systemd-resolved(8):

> * Alternatively, /etc/resolv.conf may be managed by other packages, in which
>   case systemd-resolved will read it for DNS configuration data.  In this
>   mode of operation systemd-resolved is consumer rather than provider of this
>   configuration file.
> 
> Note that the selected mode of operation for this file is detected fully
> automatically, depending on whether /etc/resolv.conf is a symlink to
> /run/systemd/resolve/resolve.conf or lists 127.0.0.53 as DNS server.

Also, any idea how /etc/resolv.conf ended up as a normal file but still owned
by systemd-resolve?

> ```
> kmille@linbox:~ ls /etc/resolv.conf 
> lrwxrwxrwx 1 root root 41 Nov 14 13:30 /etc/resolv.conf -> /var/run/systemd/resolve/stub-resolv.conf
> ```

Did you manually point it to /var/run instead of /run?

AFAIK distributions with systemd do not use /var/run.

**Comment by kmille** [comment]
> Could you clarify?
> Having it as a normal file seems like a valid configuration.

True. But I exepcted it to be a symlink to the systemd-resolved stub.

> Also, any idea how /etc/resolv.conf ended up as a normal file but still owned by systemd-resolve?

No. Part of the "misconfiguration"...

> Did you manually point it to /var/run instead of /run?

No particular reason. The [docs](https://wiki.archlinux.org/title/Systemd-resolved) also link to `/run`. I will change it on my side.

:v:

**Comment by kmk3** [comment]
> Is [#6296](https://github.com/netblue30/firejail/issues/6296) the same error?

Yes, good catch.
